python/cuml/naive_bayes/naive_bayes.py

```diff
@@ -58,19 +58,48 @@ def count_features_coo_kernel(float_dtype, int_dtype):
       int row = rows[i];
       int col = cols[i];
       {0} val = vals[i];
-
+      {1} label = labels[row];
+      unsigned out_idx = (col * n_classes) + label;
       if(has_weights)
         val *= weights[i];
 
       if(square) val *= val;
-      {1} label = labels[row];
-      atomicAdd(out + ((col * n_classes) + label), val);
+      atomicAdd(out + out_idx, val);
     }'''
 
     return cuda_kernel_factory(kernel_str, (float_dtype, int_dtype),
                                "count_features_coo")
 
 
+def count_features_categorical_coo_kernel(float_dtype, int_dtype):
+    """
+    A simple reduction kernel that takes in a sparse (COO) array
+    of features and computes the sum (or sum squared) for each class
+    label
+    """
+
+    kernel_str = r'''(int *out_rows, int *out_cols,
+                    int *rows, int *cols,
+                    {0} *vals, int nnz,
+                    int n_classes, int n_cols,
+                    {1} *labels) {
+
+      int i = blockIdx.x * blockDim.x + threadIdx.x;
+
+      if(i >= nnz) return;
+
+      int row = rows[i];
+      int col = cols[i];
+      {0} val = vals[i];
+      {1} label = labels[row];
+      out_rows[i] = col + n_cols * label;
+      out_cols[i] = val;
```

Comment:
This looks like it's been reduced to an element-wise operation (e.g. 1:1 input to output). If we can use Cupy's [Elementwise Kernel](https://docs.cupy.dev/en/stable/reference/generated/cupy.ElementwiseKernel.html#cupy.ElementwiseKernel) for this then it'll simplify the code here quite a bit. It's also nice that we would be able to say (perhaps in the blog?) that we can support all of the naive bayes algorithms with only 2 custom kernels- one for counting features and another for counting classes.

Reply:
True nice suggestion. I made the changes to use Elementwise kernel